### DISCUSSION TITLE
[convert shell script to .ps1 project] Make powershell v8 great again

### DISCUSSION THREAD

**Discussion by kasini3000** [PowerShell/PowerShell#16393] [Enhancements]
Q: How to make powershell v8 great again?
A:
The sql project team has an [import tool] that translates sql into t-sql compatible statements. Powershell should also have a similar migration tool.
Translate the shell script into a powershell script. Can give powershell more advantages.

I think such a project is not difficult. 

What are the best practices?
Make it into a powershell module, or an extension of vscode?

Is anyone willing to cooperate? Welcome to discuss.

------

### Convention:
.win.ps1 #win
.linux.ps1 #linux
.osx.ps1 #mac
.arm.ps1 #arm cpu

### Translation Grammar
* Translate for, foreach:
* Translate if etc.

### Translation command:

------

Welcome to use my project.

[kasini3000](https://gitee.com/chuanjiao10/kasini3000)    win,linux devops automation batch script framework.(It is similar to ansible)

**Comment by doctordns** [comment]
Can you explain a bit more what you want?

**Comment by kasini3000** [comment]
hi doctordns: 
I have searched [convert shell to powershell tool
], I found that some people had this need many years ago.

As powershell enters linux, it is gradually recognized by linux users. With the release of kasini3000, convert
  The demand for bash to powershell is gradually increasing.


My initial thoughts:
It should be a combination of a set of powershell scripts,
It should be a parser for static bash script scripts,
It works by matching keywords, matching to key1 in the hash table, and replacing it with key2
It can convert **for-in-do-done** into **foeach ($a in $b) ｛｝**
convert **if-then** into **if () {} else{}**
convert shell syntax, shell commands...

Is anyone interested in this matter?
Is it possible to create an empty project first and put in ideas?

**Comment by rhubarb-geek-nz** [comment]
Other than trivial scripts, I am not sure that it can be done as you suggest with simple replacement. Part of the problem is that the way shell scripts work is that variables are expanded at runtime, so you don't know the contents until runtime. So you would need to do the parsing at runtime, not as a static activity. In effect you would be writing a UNIX shell interpreter in .NET/PowerShell, so that it reads shell scripts at runtime, parses according to shell script rules and then invokes each of the child processes using PowerShell. So that certainly can be done, but is not what you were describing.

For example the only way do deal with the following shell script is runtime parsing, not static parsing...
```
#!/bin/sh
A=$(cat foo)
B=$(cat bar)
$A $B
```
You don't know the contents of the files foo and bar until runtime, so you have no idea ahead of time what $A and $B will expand to. When they do expand, you need to parse the resulting text as a shell script.

The second problem is that shell scripts depend on the programs that they invoke to determine the behaviour, shell has very few internal commands, even "[" is actually "/usr/bin/[", in early days that was a symbolic link to "/usr/bin/test". So expanding
```
if [ $A = 0 ]
then
    echo it is zero
fi
```
Actually means to execute the program on the path called "[" and give it the expanded command line arguments and then abide by the status code. Now, on a Windows machine, I don't think many systems have programs called "[.exe", but you would need to provide one for that construct to work. So while a shell interpreter written in PowerShell may run on Linux exactly the same as a existing shell interpreter written in C, running the same script using the same PowerShell shell interpreter would not work on Windows because none of the underlying commands exist.

So you could achieve what you want of a converter that will take a shell script and emits a powershell script, but the most honest result would be a single line that has

```
execute-as-bash "#!/bin/sh\nentire-contents-of-original-shell-script-with-escaping-to-encode-as-a-single-argument" $args
```
So the original script has been wrapped as a PowerShell but will be interpreted as a shell script, but that still has the problem of the underlying commands not existing in all environments.

If you were to write a UNIX shell interpreter, I would suggest that .NET C# is a better language to do that in rather than PowerShell, in the same way that bash, ksh, csh etc are written in C.

**Reply by kasini3000** [reply]
```
#!/bin/sh
A=$(cat foo)
B=$(cat bar)
$A $B
```


find keyword1 , replace to keyword2:
find string on '=' left,add $
find '#!' ,replace 'sh' or 'bash' to 'pwsh'
find 'cat' ,replace to 'get-content'
keword1 ,keyword2,store to db or hashtable.

Of course, after the conversion is completed, a little manual adjustment is required.

**Reply by rhubarb-geek-nz** [reply]
So how do you now process the line which says

```
$A $B
```

That could expand out to be any valid shell command, but at initial translation time you have no idea what it will be. For example if A contained "cat" and B contained a "hello.txt", then you would be expected to execute

```
cat hello.txt
```

However your example says you should be converting all "cat" as a command to "get-content". So you can only do that by parsing at runtime, because you don't know what the contents of foo and bar will be until you run the script.

Hence why I suggest you need the capability of something like execute-as-bash to process shell scripts at runtime

**Reply by rhubarb-geek-nz** [reply]
I suggest you may store the A and B directly in the environment variables, that is how shell script effectively processes them and allows them to be expanded. The tricky part is where shell has the concept of environment variables that are exported to child processes, so they have to be marked in someway, or else have environment variable and private environment variables where a private variable would be promoted to the main process variables when it is exported so it can be inherited by child processes.

**Reply by doctordns** [reply]
This is an interesting idea, but it seems far from trivial.  Unless you can build a good conversion tool, I am not sure of the likely take up. 

I suggest the OP develops a basic proof of concept implementation to enable more detailed discussions.

**Comment by rhubarb-geek-nz** [comment]
You would need something like execute-as-bash in order to process shell's eval command which takes a string and interprets as shell.

**Comment by rhubarb-geek-nz** [comment]
Unfortunately, once you say it is the job of the tool to convert from "cat" to "get-content" you have changed the scope of the project from the scripting language itself which has a limited vocabulary, to now being responsible for having a mapping for every command line program and mapping its name, format of arguments and calling conventions. 

In terms of shell scripting, terms like "if", "for", "while", "do", "cd" are part of the shell vocabulary. Programs such as "ls", "cat", "echo", "test", "[" are part of the operating system which can have different arguments depending on the flavour of the operating system. So while they are common, and POSIX has attempted to define common arguments they still have their own quirks based on the OS.

So to compare with an SQL convertor tool, it can do conversions based on the SQL language itself, but it can't make inferences regarding what the tables are called, what the contents are and how they should be treated. A shell language invokes programs, but they are not part of the shell language.

In practical terms, what it means is that although you can have a common scripting language, it does not change the underlying operating system and you won't have portable scripts between windows and POSIX unless the script limits itself to a restricted set of built in PowerShell operations. 

Now, perhaps that is as far as you can practically go, if an existing shell script is very basic and only makes simple calls to known programs that have an existing mapping, then that is achievable.

But to make that open ended, you need to keep on extending the set of PowerShell modules to map to all the operations and new programs you want to cover and support translation for.

**Comment by rhubarb-geek-nz** [comment]
To look at the scale of the problem do a google search for UNIX Rosetta Stone, there are many of them!

**Comment by doctordns** [comment]
This has been an interesting thread. Were such a tool to be created, it would certainly appeal to some - provided it was 100% reliable. I have concerns over how easy this would be and on the scope of resources that might be needed to deliver a rock-solid product. I suspect from an investment decision - this is probably not top of anyone's list of needed features. But if the OP can do some work and make a start, who knows where it will go. 

With that said, this tool belongs in a separate module and not inside PowerShell itself. My advice to the OP is to should open up a separate repo here on GitHub and start to code and promote the concepts and begin to recruit other contributors. 

In the meantime _this_ issue can probably be closed.

**Comment by BaptisteKrc** [comment]
Hello, it's maybe to late, but i do the exact same research as you : "convert shell to powershell tool" and i didnt find anything. I need to translate hundreads of scripts, do you work on some code ? where are you with the project ? @kasini3000

**Reply by rhubarb-geek-nz** [reply]
What is your scenario? Do you really need to convert all the scripts that I presume are tested and working to Powershell? If you are on Linux, can you not call the existing shell scripts from Powershell? If you are on Windows, can you run Linux in Docker under WSL2 to run the shell scripts unchanged?

**Reply by BaptisteKrc** [reply]
Yes, I have to convert a software from unix to windows, and there are many .sh scripts that use Fortran. So I think I have to convert all of them in .ps1 (no bat allowed). I did not know about the last thing you mention I'll check if it can work for my use.

**Reply by kasini3000** [reply]
>where are you with the project ?

This project has not started yet and I am looking forward to working on this project under psteam. Let me restate what this project does:
1 I expect it to be a static code analyzer that translates shell syntax into powershell syntax,
2 I expected it would not translate “cat” into “get-content”. This matter can be handed over to vscode, and vscode can already do this.
3 Maybe add a feature: make it convert .bat to .ps1

The generated .ps1 script can also be used both on linux and on windows. But since there are windows users who need it, I expect Microsoft's technical vice president (win project) to pay attention to and invest in this project.

**Reply by BaptisteKrc** [reply]
Dont know about psteam sorry I'm an engineering school student (not much experience)  , by the way I didn't know that VSCODE could do "translation". Bat to ps1 can be interesting too
